Merge RTEMS5 changes and rebase on a recent 7.0 base (was bcrtems5)

brendanchandler opened the pull request:
These changes start with a base of the current 7.0 branch and then merge mdavidsaver/rtems5 followed by heinz/7.0.  Hopefully that picks up all the rtems5 changes so far, and puts them on a recent base of the main 7.0 branch.

Unfortunately, I don't have everything working yet.  Compiling with RTEMS5 on pc686 works.  Compiling with RTEMS 4.10.2 on pc386 still has issues.

I'm going to try to fix the rest of the issues, but I thought I'd put this out here FYI in case anyone wants to help.

mdavidsaver committed:
Extract compiler pre-defined macros as Make variables

hjunkes committed:
Enable Epics for RTEMS5 (posix)

hjunkes committed:
Enable Epics for RTEMS5 (posix)

hjunkes committed:
No Windows, rtems without test

hjunkes committed:
Prepared RTEMS 4.10 posix with network

hjunkes committed:
Prepared RTEMS 4.10 posix with network

hjunkes committed:
Merge branch '7.0' of https://github.com/hjunkes/epics-base into 7.0

mdavidsaver committed:
WIP: disable appveyor

mdavidsaver committed:
ci: RTEMS update

mdavidsaver committed:
RTEMS: auto-detect OS_API

mdavidsaver committed:
epicsNtp: avoid deprecated/non-standard bzero()

mdavidsaver committed:
WIP: fix missing bits needed by osiNTPTime.c

mdavidsaver committed:
RTEMS5: redirect to serial via runtime

mdavidsaver committed:
RTEMS: fix NIC for testing w/ RTEMS5

Add two NICs, ne2k used by RTEMS 4.x
and e1000 used by RTEMS 5.1.
Each ignores the NIC it doesn't understand.

mdavidsaver committed:
re-drop RTEMS osdMessageQueue

hjunkes committed:
Fixed stupid typo in CONFIG.Common.RTEMS

anjohnson committed:
Install toolchain data into cfg/ directory

anjohnson committed:
Set GNU_TARGET automatically

The individual CONFIG.Common.RTEMS-$(T_A) files don't really need
to set the new GNU_TARGET variable themselves, this addition allows
the unmodified RTEMS-uC5282 target build fine (on 4.10.1 at least).

anjohnson committed:
Explain the RTEMS config variables

anjohnson committed:
Move HAVE_SOCKADDR_SA_LEN=1 for darwin targets

anjohnson committed:
Add RTEMS-pc686-qemu for use on RTEMS 5.x

RTEMS 5.x moved the PC BSP to 686. This commit makes the EPICS
target match; RTEMS-pc386 and the -qemu variant can only be built
when RTEMS_VERSION != 5, whereas RTEMS-pc686 and its -qemu version
will only build when RTEMS_VERSION == 5 (there are checks with
descriptive errors included).

makeTestFile.pl was also reformatted and modified to use exec to
run tests on non-Windows hosts as required by the Perl test harness.

anjohnson committed:
Clean up CONFIG.Common.RTEMS

anjohnson committed:
Enable RTEMS testing in modules/database/test/std/link

hjunkes committed:
cleanup netconfig

hjunkes committed:
forgot to add, add definition for GNU_TARGET

hjunkes committed:
MQ name creation changed

anjohnson committed:
Set RTEMS_TARGET for Travis builds

anjohnson committed:
Simplify RTEMS OP_SYS_CFLAGS settings

mdavidsaver committed:
rtems5: fix linking of dependent modules, need -lCom after -lrtemsCom

mdavidsaver committed:
rtems5: correct ifreq_size()

mdavidsaver committed:
restore $(API_CPPFLAGS)

Commit:
changed pvAccess url, save adaption to old network stack

Commit:
Link pvData module to mdavidsaver

Commit:
Add LEGACY_STACK_SUPPORT and telnet client

Commit:
Add posix cpp flags to RTEMS config

hjunkes committed:
Cosmetics and some typos

hjunkes committed:
Add dhcp option cmdline (129)

Commit:
Merge remote-tracking branch 'mdavidsaver/rtems5' into bcrtems5

Commit:
Merge remote-tracking branch 'hjunkes/7.0' into bcrtems5

Commit:
Fix merge issues

Commit:
Resolve some merge issues.

There are changes that made it through the merge, but I think are no
longer needed.

osdNTPGet doesn't appear to be used anywhere, I think its use was
removed but the definition of the function made it through the merge.

The include of sysinit.h existed on one branch, but not on another and
it caused a compile error, so I removed it for now.

rtems_bsd_set_vprintf_handler failed on RTEMS 4.10.2 so was commented out.

anjohnson commented on configure/CONFIG_SITE:
```diff
@@ -103,7 +103,7 @@
 #  Definitions in configure/os/CONFIG_SITE.<host>.Common
 #  may override this setting.
-CROSS_COMPILER_TARGET_ARCHS=
+CROSS_COMPILER_TARGET_ARCHS=RTEMS-qoriq_e500
```
Please remove this change from your PR. You can put that definition in a configure/CONFIG_SITE.local file in your working tree where it will have the same effect for you, but putting this in the PR breaks all our CI builds.

brendanchandler replied:
Removed.

anjohnson commented on .travis.yml:
```diff
@@ -0,0 +1,93 @@
+# .travis.yml for use with EPICS Base ci-scripts
```
The `.travis.yml` file has been removed from the 7.0 branch of Base, we've switched from Travis (which is no longer free) to using GitHub Actions instead, so please remove this file again.

brendanchandler replied:
Removed

anjohnson commented on .github/workflows/ci-scripts-build.yml:
```diff
@@ -68,17 +70,26 @@ jobs:
+            cmp: gcc
+            configuration: default
+            rtems: "5"
+            rtems_target: RTEMS-pc686-qemu
```
Please add this line to name this particular matrix entry:
```yaml
            name: "Ub-20 gcc-9 + RT-5.1"
```

brendanchandler replied:
added

anjohnson reviewed (commented):
A few fixes to be getting started with, making these should turn the non-RTEMS jobs green on both GitHub and Appveyor.

Commit:
Fixes for CI mechanics

- Removed .travis.yml since we're no longer using travis.
- Named the rtems pc686 target in github actions
- Cleaned up CONFIG_SITE.Common.RTEMS

brendanchandler commented on .gitmodules:
```diff
@@ -1,10 +1,10 @@
     path = modules/pvData
-    url = https://github.com/epics-base/pvDataCPP
+    url = https://github.com/mdavidsaver/pvDataCPP
     branch = master
 [submodule "modules/pvAccess"]
     path = modules/pvAccess
-    url = https://github.com/epics-base/pvAccessCPP
+    url = https://github.com/mdavidsaver/pvAccessCPP
```
Should we keep this as epics-base?

mdavidsaver replied:
If the submodule revisions are now in main repo (I think there are), then these can be switched back.

brendanchandler commented on configure/os/CONFIG_SITE.Common.RTEMS:
```diff
@@ -16,8 +16,8 @@ RTEMS_VERSION = 5
-#RTEMS_VERSION = 5
-#RTEMS_BASE = /home/h1/DBG/rtems
+RTEMS_VERSION = 5
+RTEMS_BASE = /opt/RTEMS/qoriq/rtems/$(RTEMS_VERSION)
```
Does it make sense to set RTEMS_VERSION and RTEMS_BASE here?  I should comment out these lines right?

mdavidsaver replied:
It has been on my TODO list to see if these could simply be moved to `configure/CONFIG_SITE.local`.

brendanchandler replied:
I've been setting them in CONFIG_SITE.local in my local repo without issue so far.

anjohnson replied:
CONFIG_SITE.local gets pulled in by CONFIG_SITE which is the first such file to be included so yes, these can be set there, but will be overridden by any values that later get set in this file. Historically my APS Jenkins jobs have never had to configure them because the repo has always defaulted to our settings, but I've just fixed that (I think).

mdavidsaver replied:
Does this mean that the "defaults" in `configure/os/CONFIG_SITE.Common.RTEMS` can be commented out now?

anjohnson replied:
Yes, I updated my Jenkins job configurations to set them itself.

Commit:
Remove tab from ci-scripts-build.yml

anjohnson commented on configure/CONFIG_COMMON:
```diff
@@ -306,7 +307,7 @@ LDLIBS = $(POSIX_LDLIBS) $(ARCH_DEP_LDLIBS) $(DEBUG_LDLIBS) $(OP_SYS_LDLIBS)\
  $(DEBUG_CPPFLAGS) $(WARN_CPPFLAGS) $(BASE_CPPFLAGS) $(TARGET_CPPFLAGS)\
  $(USR_CPPFLAGS) $(CMD_CPPFLAGS) $(ARCH_DEP_CPPFLAGS) $(OP_SYS_CPPFLAGS)\
- $(OP_SYS_INCLUDE_CPPFLAGS) $(CODE_CPPFLAGS) $(API_CPPFLAGS)
+ $(OP_SYS_INCLUDE_CPPFLAGS) $(CODE_CPPFLAGS)
```
I'm pretty sure this change is what has broken the Windows dynamic builds, the `-DBUILDING_LIBCOM_API` and similar flags get put into the `API_CPPFLAGS` variable in `configure/RULES_TARGET` so please restore it, which should add yet more green to the board!

Commit:
Reintroduce API_CPPFLAGS argument to fix Windows builds

Commit:
Make pvData and pvAccess submodules use epics-base

Previously they used mdavidsaver's branch in order to pick up some
commits that hadn't made it into mainline yet.

Commit:
Move rtems_netconfig.c and libbsd_netconfig.c out of OS specific dir

These files aren't specific to the version of RTEMS we're using, but
rather whether networking is enabled or not.  Putting them in an RTEMS
4/5 specific directory only allowed successful build when networking
was disabled on RTEMS 5 or enabled on RTEMS 4, otherwise make couldn't
find the .c files.

Commit:
Let OS_API be set by configure.c again, clean up LDLIBS

The merge re-introduced setting OS_API in CONFIG.Common.RTEMS, masking
over the value set by configure.c.  This change removes the setting of
OS_API so it gets set automatically instead.

It also cleans up some of the LDLIBS flags, similar to how they were cleaned
up on mdavidsaver/rtems5 branch, but conflicts with hjunkes/7.0 branch
got things out of whack.

Commit:
Remove libbsd_netconfig.c in favor of combined netconfig file

This file was re-introduced by the merge, but is mean to be deleted.

brendanchandler commented:
Closing pull request in favor of my other pull request (brendanchandler/rtems5), which rebases all the changes from hjunkes, anjohnson, and mdavidsaver rather than merging them (as requested by the core developers).
